File: MyVoice/OverlayPositioner.swift
import AppKit

/// Calculates overlay window position relative to the caret.
enum OverlayPositioner {
    private static let verticalOffset: CGFloat = 4

    static func position(for caretRect: CGRect, overlaySize: NSSize) -> NSPoint {
        // Position at upper-left of the caret
        // Use fixed offsets from caret origin, ignore reported caret height
        var x = caretRect.origin.x - overlaySize.width + 8  // slightly overlap horizontally
        var y = caretRect.origin.y - 12                        // align with top of cursor

        // Clamp to screen bounds
        if let screen = NSScreen.main {
            let screenFrame = screen.visibleFrame

            // Don't go off left edge — flip to right of caret
            if x < screenFrame.origin.x {
                x = max(screenFrame.origin.x, caretRect.origin.x + verticalOffset)
            }

            // Don't go off right edge
            if x + overlaySize.width > screenFrame.maxX {
                x = screenFrame.maxX - overlaySize.width
            }

            // Don't go off top edge — flip below caret
            if y + overlaySize.height > screenFrame.maxY {
                y = caretRect.origin.y - overlaySize.height - verticalOffset
            }

            // Don't go off bottom edge
            if y < screenFrame.origin.y {
                y = screenFrame.origin.y
            }
        }

        return NSPoint(x: x, y: y)
    }
}

File: windows/src/MyVoice.Windows/Core/OverlayPositioner.cs
using System.Drawing;

namespace MyVoice.Windows.Core;

/// <summary>
/// Where the recording overlay goes (the Mac's OverlayPositioner, in Windows screen coordinates: origin top-left,
/// y downward). Upper-left of the caret, its bottom edge on the caret's line, kept inside the work area of the
/// caret's monitor — which can sit at negative x or y.
/// </summary>
public static class OverlayPositioner
{
    /// <param name="caret">Caret (or mouse) rectangle in physical screen pixels.</param>
    /// <param name="overlay">Overlay size in physical pixels.</param>
    /// <param name="workArea">Work area of the caret's monitor (without the taskbar).</param>
    /// <param name="scale">That monitor's DPI / 96: the Mac's offsets are in 96-DPI pixels.</param>
    public static Point Position(Rectangle caret, Size overlay, Rectangle workArea, double scale = 1)
    {
        int Px(double points) => (int)Math.Round(points * scale);
        var x = caret.Left - overlay.Width + Px(8); // overlaps the caret slightly, like the Mac
        var y = caret.Top + Px(12) - overlay.Height;

        if (x < workArea.Left) x = Math.Max(workArea.Left, caret.Left + Px(4));   // off the left edge: right of the caret
        if (x + overlay.Width > workArea.Right) x = workArea.Right - overlay.Width;
        if (y < workArea.Top) y = caret.Bottom + Px(4);                           // off the top: below the caret
        if (y + overlay.Height > workArea.Bottom) y = workArea.Bottom - overlay.Height;

        return new Point(Math.Max(x, workArea.Left), Math.Max(y, workArea.Top));
    }
}
